pkg/fqdn/name_manager_checkpoint.go

```diff
@@ -0,0 +1,99 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package fqdn
+
+import (
+	"cmp"
+	"context"
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+	"slices"
+
+	"github.com/google/renameio/v2"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/policy/api"
+	"github.com/cilium/cilium/pkg/time"
+	"github.com/cilium/cilium/pkg/trigger"
+)
+
+const (
+	checkpointFile = "fqdn-name-manager-selectors.json"
+
+	checkpointController      = "fqdn-selector-checkpointing"
+	checkpointTriggerInterval = 10 * time.Second
+)
+
+// serializedSelector is the schema of the serialized selectors on disk
+type serializedSelector struct {
+	Regex    string           `json:"re"`
+	Selector api.FQDNSelector `json:"sel"`
+}
+
+// setupSelectorCheckpointController returns a new trigger, which when invoked,
+// will write all selectors known to the NameManager to disk. This allows Cilium 1.16+
+// (which has a new way of managing identities for FQDN selectors) to restore the
+// IPCache with selector-based identities and thus allow for a dropless upgrade.
```

Comment:
```suggestion
// IPCache with selector-based identities and thus allows for a dropless upgrade.
```

Also, in commit title: “compatiblity” (I fixed it in the PR title).